### PR TITLE
ONYX-11118 - FetchAudienceValue class

### PR DESCRIPTION
**Pull request by sashaCher**
### What does this PR do?

Introduces `FetchAudienceValue` class responsible for fetching `audience` variable of `author-jwt`.

#### Change log
- [ ] The CHANGELOG has been updated, or
- [X] This PR does not include user-facing changes and doesn't require a CHANGELOG update

#### Test coverage
- [X] This PR includes new unit and integration tests to go with the code changes, or
- [ ] The changes in this PR do not require tests

#### Documentation
- [ ] Docs (e.g. `README`s) were updated in this PR, and/or there is a follow-on issue to update docs, or
- [X] This PR does not require updating any documentation

#### API Changes
- [ ] The [OpenAPI spec](https://github.com/cyberark/conjur-openapi-spec) has been updated to meet new API changes (or an issue has been opened), or
- [X] The changes in this PR do not affect the Conjur API

### CODE REVIEW

File: app/domain/authentication/authn_jwt/validate_and_decode/fetch_audience_value.rb [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        end
+
+        def audience_secret_value_is_not_empty
+          raise Errors::Authentication::AuthnJwt::AudienceValueIsEmpty if audience_secret_value.blank?
+        end
+
+        # https://datatracker.ietf.org/doc/html/rfc7519#section-2 => StringOrURI
+        # any value containing a ":" character MUST be a URI [RFC3986]
+        def audience_secret_value_is_stringoruri
+          begin
```
[the remark this answers is not in the document]
Comment (tzheleznyak) [reply app/domain/authentication/authn_jwt/validate_and_decode/fetch_audience_value.rb:70]: @sashaCher Fix this important error

File: app/domain/authentication/authn_jwt/validate_and_decode/fetch_audience_value.rb [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+            URI::RFC3986_PARSER.parse(audience_secret_value) if audience_secret_value.include?(":")
+          rescue => e
+            raise Errors::Authentication::AuthnJwt::AudienceValueIsNotURI.new(e.inspect)
```
Comment (tzheleznyak) [inline app/domain/authentication/authn_jwt/validate_and_decode/fetch_audience_value.rb:73]:
```suggestion
            raise Errors::Authentication::AuthnJwt::AudienceValueIsNotURI, e.inspect
```

File: app/domain/authentication/authn_jwt/validate_and_decode/fetch_audience_value.rb [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+      # Fetch and validate the audience from the JWT authenticator policy
+      FetchAudienceValue = CommandClass.new(
+        dependencies: {
+          resource_class: ::Resource,
+          fetch_required_secrets: ::Conjur::FetchRequiredSecrets.new,
+          logger: Rails.logger
+        },
+        inputs: %i[authentication_parameters]
+      ) do
+        def call
```
Comment (nessiLahav) [inline app/domain/authentication/authn_jwt/validate_and_decode/fetch_audience_value.rb:15]: add info message of the final audience value

File: app/domain/authentication/authn_jwt/validate_and_decode/fetch_audience_value.rb [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+          @audience_resource ||= @resource_class[audience_resource_id]
+        end
+
+        def audience_resource_id
+          @audience_resource_id ||= "#{@authentication_parameters.authn_jwt_variable_id_prefix}/#{AUDIENCE_RESOURCE_NAME}"
+        end
+        
+        def empty_audience_value
+          @logger.debug(LogMessages::Authentication::AuthnJwt::FetchingAudienceValue.new)
+          @empty_audience ||= ""
```
Comment (tzheleznyak) [inline app/domain/authentication/authn_jwt/validate_and_decode/fetch_audience_value.rb:43]: Why memoazing empty string

File: app/domain/authentication/authn_jwt/validate_and_decode/fetch_audience_value.rb [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        def audience_secret_value
+          @audience_secret_value ||= audience_required_secret[audience_resource_id]
+        end
+        
+        def audience_required_secret
+          @audience_required_secret ||= @fetch_required_secrets.call(resource_ids: [audience_resource_id])
+        end
+
+        def validate_audience_secret_value
+          audience_secret_value_is_not_empty
```
Comment (tzheleznyak) [inline app/domain/authentication/authn_jwt/validate_and_decode/fetch_audience_value.rb:59]: I think we should rename this two functions and move them to call

File: app/domain/authentication/authn_jwt/validate_and_decode/fetch_audience_value.rb [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+          audience_secret_value_is_stringoruri
+        end
+
+        def audience_secret_value_is_not_empty
+          raise Errors::Authentication::AuthnJwt::AudienceValueIsEmpty if audience_secret_value.blank?
+        end
+
+        # https://datatracker.ietf.org/doc/html/rfc7519#section-2 => StringOrURI
+        # any value containing a ":" character MUST be a URI [RFC3986]
+        def audience_secret_value_is_stringoruri
```
Comment (nessiLahav) [inline app/domain/authentication/authn_jwt/validate_and_decode/fetch_audience_value.rb:69]: the 3rd do not know to perform this use case ?

File: app/domain/authentication/authn_jwt/validate_and_decode/fetch_audience_value.rb [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+        def audience_required_secret
+          @audience_required_secret ||= @fetch_required_secrets.call(resource_ids: [audience_resource_id])
+        end
+
+        def validate_audience_secret_value
+          audience_secret_value_is_not_empty
+          audience_secret_value_is_stringoruri
+        end
+
+        def audience_secret_value_is_not_empty
```
Comment (nessiLahav) [inline app/domain/authentication/authn_jwt/validate_and_decode/fetch_audience_value.rb:63]: maybe it should changed to `validate_audience_secret_has_value` ?

**Review by nessiLahav — CHANGES_REQUESTED** [review]
few comments

File: app/domain/authentication/authn_jwt/validate_and_decode/fetch_audience_value.rb (hunk 6 above)
Comment (sashaCher) [reply app/domain/authentication/authn_jwt/validate_and_decode/fetch_audience_value.rb:69]: `jet` does not do this check

File: app/domain/authentication/authn_jwt/validate_and_decode/fetch_audience_value.rb (hunk 2 above)
Comment (sashaCher) [reply app/domain/authentication/authn_jwt/validate_and_decode/fetch_audience_value.rb:73]: Parenthesis are always working.
`,` coma only for the cases were there's only one parameter
So I prefer to keep it this way

File: app/domain/authentication/authn_jwt/validate_and_decode/fetch_audience_value.rb (hunk 5 above)
Comment (sashaCher) [reply app/domain/authentication/authn_jwt/validate_and_decode/fetch_audience_value.rb:59]: Only if @nessiLahav agrees

File: app/domain/authentication/authn_jwt/validate_and_decode/fetch_audience_value.rb (hunk 3 above)
Comment (sashaCher) [reply app/domain/authentication/authn_jwt/validate_and_decode/fetch_audience_value.rb:15]: Done

File: app/domain/authentication/authn_jwt/validate_and_decode/fetch_audience_value.rb (hunk 4 above)
Comment (sashaCher) [reply app/domain/authentication/authn_jwt/validate_and_decode/fetch_audience_value.rb:43]: Done

File: app/domain/authentication/authn_jwt/validate_and_decode/fetch_audience_value.rb (hunk 7 above)
Comment (sashaCher) [reply app/domain/authentication/authn_jwt/validate_and_decode/fetch_audience_value.rb:63]: Done

File: app/domain/authentication/authn_jwt/validate_and_decode/fetch_audience_value.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (sashaCher) [reply app/domain/authentication/authn_jwt/validate_and_decode/fetch_audience_value.rb:70]: Done

File: app/domain/authentication/authn_jwt/validate_and_decode/fetch_audience_value.rb (hunk 2 above)
Comment (sashaCher) [reply app/domain/authentication/authn_jwt/validate_and_decode/fetch_audience_value.rb:73]: Done

Comment (sashaCher) [reply app/domain/authentication/authn_jwt/validate_and_decode/fetch_audience_value.rb:73]: Parenthesis are always working.
`,` coma only for the cases were there's only one parameter.
So I prefer to keep it this way
In addition raise does not evaluate `e.inspect` into placeholder in the `AudienceValueIsNotURI` message.

File: app/domain/authentication/authn_jwt/validate_and_decode/fetch_audience_value.rb [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+          @audience_resource ||= @resource_class[audience_resource_id]
+        end
+
+        def audience_resource_id
+          @audience_resource_id ||= "#{@authentication_parameters.authn_jwt_variable_id_prefix}/#{AUDIENCE_RESOURCE_NAME}"
+        end
+        
+        def empty_audience_value
+          @logger.debug(LogMessages::Authentication::AuthnJwt::FetchingAudienceValue.new)
+          String.new
```
[the remark this answers is not in the document]
Comment (nessiLahav) [reply app/domain/authentication/authn_jwt/validate_and_decode/fetch_audience_value.rb:46]: please fix it

File: app/domain/authentication/authn_jwt/validate_and_decode/fetch_audience_value.rb (hunk 6 above)
Comment (nessiLahav) [reply app/domain/authentication/authn_jwt/validate_and_decode/fetch_audience_value.rb:72]: i think this check is redundant, why it is important to us what is the audience value  ?

File: app/domain/authentication/authn_jwt/validate_and_decode/fetch_audience_value.rb (hunk 8 above)
[the remark this answers is not in the document]
Comment (sashaCher) [reply app/domain/authentication/authn_jwt/validate_and_decode/fetch_audience_value.rb]: Done

File: app/domain/authentication/authn_jwt/validate_and_decode/fetch_audience_value.rb (hunk 6 above)
Comment (sashaCher) [reply app/domain/authentication/authn_jwt/validate_and_decode/fetch_audience_value.rb]: Agreed

**Review by nessiLahav — APPROVED** [review]
LGTM

**Pull request merged**